File: Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        ans = []
        
        # depth first search for all possibilities
        def dfs(s='', left=0, right=0):
            if len(s) == 2 * n:
                ans.append(s)
            else:
                if left > right:
                    dfs(s+')', left, right+1)
                if left < n:
                    dfs(s+'(', left+1, right)
        
        dfs()
        
        return ans

File: C++/Generate Parentheses.cpp
class Solution {
private:
    void dfs(vector<string>& ans, string str, int left, int right) {
        if(left == 0 and right == 0)
            ans.push_back(str);
        
        if(left > 0)
            dfs(ans, str + "(", left - 1, right);
        if(right > left)
            dfs(ans, str + ")", left, right - 1);
    }
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;

        // recursively create string
        dfs(ans, "", n, n);
        
        return ans;
    }
};